tutorials/src/bin/basic-tutorial-6.rs

```diff
@@ -0,0 +1,162 @@
+        } else if pad_template.direction() == gst::PadDirection::Sink {
+            println!("  SINK template: '{}'", pad_template.name_template());
+        } else {
+            println!("  UNKNOWN!!! template: '{}'", pad_template.name_template());
+        }
+
+        if pad_template.presence() == gst::PadPresence::Always {
+            println!("  Availability: Always");
+        } else if pad_template.presence() == gst::PadPresence::Sometimes {
+            println!("  Availability: Sometimes");
+        } else if pad_template.presence() == gst::PadPresence::Request {
+            println!("  Availability: On request");
+        } else {
+            println!("  Availability: UNKNOWN!!!");
+        }
+
+        let caps = pad_template.get_caps();
+        println!("  Capabilities:");
+        print_caps(caps, "    ");
+    }
+}
+
+fn print_pad_capabilities(element: &gst::Element, pad_name: &str) {
+    let pad = element.get_static_pad(pad_name)
+        .expect("Could not retrieve pad");
+
+    println!("Caps for the {} pad:", pad_name);
+    match pad.get_current_caps() {
+      Some(caps) => {
+          print_caps(caps, "      ");
+      },
+      None => {
+          let caps = pad.query_caps(None)
+              .expect("Failed to query caps on pad");
+          print_caps(caps, "      ");
+      }
+    }
+}
+
+fn main() {
+    // Initialize GStreamer
+    gst::init().unwrap();
+
+    // Create the element factories
+    let source_factory = gst::ElementFactory::find("audiotestsrc")
+        .expect("Failed to create audiotestsrc factory.");
+    let sink_factory = gst::ElementFactory::find("autoaudiosink")
+        .expect("Failed to create autoaudiosink factory.");
+
+    // Print information about the pad templates of these factories
+    print_pad_template_information(&source_factory);
+    print_pad_template_information(&sink_factory);
+
+    // Ask the factories to instantiate actual elements
+    let source = source_factory.create("source")
+        .expect("Failed to create source element");
+    let sink = sink_factory.create("sink")
+        .expect("Failed to create sinke element");
+
+    // Create the empty pipeline
+    let pipeline = gst::Pipeline::new("test-pipeline");
+
+    pipeline.add_many(&[&source, &sink]).unwrap();
+    source.link(&sink).expect("Elements could not be linked.");
+
+    // Print initial negotiated caps (in NULL state)
+    println!("In NULL state:");
+    print_pad_capabilities(&sink, "sink");
+
+    // Start playing
+    let ret = pipeline.set_state(gst::State::Playing);
+    if ret == gst::StateChangeReturn::Failure {
+        eprintln!("Unable to set the pipeline to the playing state (check the bus for error messages).")
+    }
+
+    // Wait until error, EOS or State Change
+    let bus = pipeline.get_bus().unwrap();
+
+    let mut terminate = false;
+
+    while !terminate {
+        // FIXME: use timed_pop_filtered with GST_MESSAGE_ERROR | GST_MESSAGE_EOS | GST_MESSAGE_STATE_CHANGED
```

Comment:
The MessageType enum is not even in the bindings :) it's broken somewhat because it's flags but not really. I'd like to not expose this API. Matching on the MessageView seems good enough IMHO